File: packages/javascript/src/providers.ts
import { getConfig } from "./config.js";

export interface Message {
  role: string;
  content?: string | null;
  tool_calls?: ToolCall[];
  tool_call_id?: string;
}

export interface ToolCall {
  id: string;
  name: string;
  arguments: Record<string, unknown>;
}

export interface ProviderResponse {
  role: string;
  content?: string | null;
  tool_calls: ToolCall[];
  finish_reason?: string;
  usage?: Record<string, unknown>;
}

export interface Provider {
  name: string;
  chat(messages: Message[], tools: ToolDefinition[], model: string): Promise<ProviderResponse>;
}

export interface ToolDefinition {
  name: string;
  description: string;
  parameters: Record<string, unknown>;
}

export class OpenAIProvider implements Provider {
  name = "openai";

  async chat(messages: Message[], tools: ToolDefinition[], model: string): Promise<ProviderResponse> {
    const apiKey = getConfig().apiKeys.openai;
    if (!apiKey) throw new Error("OpenAI API key not configured");

    const body: Record<string, unknown> = {
      model,
      messages: messages.map((m) => {
        const entry: Record<string, unknown> = { role: m.role, content: m.content };
        if (m.tool_calls) {
          entry.tool_calls = m.tool_calls.map((tc) => ({
            id: tc.id,
            type: "function",
            function: { name: tc.name, arguments: JSON.stringify(tc.arguments) },
          }));
        }
        if (m.tool_call_id) entry.tool_call_id = m.tool_call_id;
        return entry;
      }),
    };

    if (tools.length) {
      body.tools = tools.map((t) => ({
        type: "function",
        function: { name: t.name, description: t.description, parameters: t.parameters },
      }));
    }

    const response = await fetch("https://api.openai.com/v1/chat/completions", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });

    const data = await response.json();
    if (!response.ok) throw new Error(data.error?.message || `OpenAI error ${response.status}`);

    const choice = data.choices[0];
    const message = choice.message;
    const tool_calls = (message.tool_calls || []).map((tc: { id: string; function: { name: string; arguments: string } }) => ({
      id: tc.id,
      name: tc.function.name,
      arguments: JSON.parse(tc.function.arguments),
    }));

    return {
      role: "assistant",
      content: message.content,
      tool_calls,
      finish_reason: choice.finish_reason,
      usage: data.usage,
    };
  }
}

export class AnthropicProvider implements Provider {
  name = "anthropic";

  async chat(messages: Message[], tools: ToolDefinition[], model: string): Promise<ProviderResponse> {
    const apiKey = getConfig().apiKeys.anthropic;
    if (!apiKey) throw new Error("Anthropic API key not configured");

    const systemParts: string[] = [];
    const conversation: Message[] = [];
    for (const m of messages) {
      if (m.role === "system") systemParts.push(m.content || "");
      else conversation.push(m);
    }

    const body: Record<string, unknown> = {
      model,
      max_tokens: 4096,
      messages: conversation.map((m) => {
        const role = m.role === "tool" ? "user" : m.role;
        if (m.tool_calls) {
          return {
            role,
            content: m.tool_calls.map((tc) => ({
              type: "tool_use",
              id: tc.id,
              name: tc.name,
              input: tc.arguments,
            })),
          };
        }
        if (m.tool_call_id) {
          return {
            role,
            content: [{ type: "tool_result", tool_use_id: m.tool_call_id, content: m.content || "" }],
          };
        }
        return { role, content: m.content || "" };
      }),
    };

    if (systemParts.length) body.system = systemParts.join("\n\n");
    if (tools.length) {
      body.tools = tools.map((t) => ({
        name: t.name,
        description: t.description,
        input_schema: t.parameters,
      }));
    }

    const response = await fetch("https://api.anthropic.com/v1/messages", {
      method: "POST",
      headers: {
        "x-api-key": apiKey,
        "anthropic-version": "2023-06-01",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });

    const data = await response.json();
    if (!response.ok) throw new Error(data.error?.message || `Anthropic error ${response.status}`);

    const textParts: string[] = [];
    const tool_calls: ToolCall[] = [];
    for (const block of data.content || []) {
      if (block.type === "text") textParts.push(block.text);
      else if (block.type === "tool_use") {
        tool_calls.push({ id: block.id, name: block.name, arguments: block.input || {} });
      }
    }

    return {
      role: "assistant",
      content: textParts.join("\n") || null,
      tool_calls,
      finish_reason: data.stop_reason,
      usage: data.usage,
    };
  }
}

const providers: Record<string, Provider> = {
  openai: new OpenAIProvider(),
  anthropic: new AnthropicProvider(),
};

export function getProvider(name: string): Provider {
  const provider = providers[name];
  if (!provider) throw new Error(`Unknown provider: ${name}`);
  return provider;
}

File: packages/python/rails_ai_build/providers.py
"""Model provider abstractions."""

from __future__ import annotations

import json
from abc import ABC, abstractmethod
from typing import Any, Optional

import httpx

from rails_ai_build.config import get_config


class BaseProvider(ABC):
    name: str = "base"

    def __init__(self, api_key: Optional[str] = None):
        self.api_key = api_key or get_config().api_keys.get(self.name, "")

    @abstractmethod
    def chat(
        self,
        messages: list[dict],
        tools: list[dict],
        model: str,
        **kwargs,
    ) -> dict[str, Any]:
        ...


class OpenAIProvider(BaseProvider):
    name = "openai"
    BASE_URL = "https://api.openai.com/v1"

    def chat(self, messages, tools, model, **kwargs):
        payload: dict[str, Any] = {
            "model": model,
            "messages": self._format_messages(messages),
        }
        if tools:
            payload["tools"] = [
                {
                    "type": "function",
                    "function": {
                        "name": t["name"],
                        "description": t["description"],
                        "parameters": t["parameters"],
                    },
                }
                for t in tools
            ]

        with httpx.Client(timeout=120) as client:
            response = client.post(
                f"{self.BASE_URL}/chat/completions",
                headers={"Authorization": f"Bearer {self.api_key}"},
                json=payload,
            )
            response.raise_for_status()
            return self._parse(response.json())

    def _format_messages(self, messages):
        result = []
        for msg in messages:
            entry: dict[str, Any] = {"role": msg["role"], "content": msg.get("content")}
            if msg.get("tool_calls"):
                entry["tool_calls"] = [
                    {
                        "id": tc["id"],
                        "type": "function",
                        "function": {
                            "name": tc["name"],
                            "arguments": json.dumps(tc["arguments"]),
                        },
                    }
                    for tc in msg["tool_calls"]
                ]
            if msg.get("tool_call_id"):
                entry["tool_call_id"] = msg["tool_call_id"]
            result.append(entry)
        return result

    def _parse(self, body):
        choice = body["choices"][0]
        message = choice["message"]
        tool_calls = []
        for tc in message.get("tool_calls") or []:
            tool_calls.append(
                {
                    "id": tc["id"],
                    "name": tc["function"]["name"],
                    "arguments": json.loads(tc["function"]["arguments"]),
                }
            )
        return {
            "role": "assistant",
            "content": message.get("content"),
            "tool_calls": tool_calls,
            "finish_reason": choice.get("finish_reason"),
            "usage": body.get("usage"),
        }


class AnthropicProvider(BaseProvider):
    name = "anthropic"
    BASE_URL = "https://api.anthropic.com/v1"
    VERSION = "2023-06-01"

    def chat(self, messages, tools, model, **kwargs):
        system, conversation = self._split_system(messages)
        payload: dict[str, Any] = {
            "model": model,
            "max_tokens": kwargs.get("max_tokens", 4096),
            "messages": self._format_messages(conversation),
        }
        if system:
            payload["system"] = system
        if tools:
            payload["tools"] = [
                {
                    "name": t["name"],
                    "description": t["description"],
                    "input_schema": t["parameters"],
                }
                for t in tools
            ]

        with httpx.Client(timeout=120) as client:
            response = client.post(
                f"{self.BASE_URL}/messages",
                headers={
                    "x-api-key": self.api_key,
                    "anthropic-version": self.VERSION,
                },
                json=payload,
            )
            response.raise_for_status()
            return self._parse(response.json())

    def _split_system(self, messages):
        system_parts, conversation = [], []
        for msg in messages:
            if msg["role"] == "system":
                system_parts.append(msg["content"])
            else:
                conversation.append(msg)
        return "\n\n".join(system_parts) or None, conversation

    def _format_messages(self, messages):
        result = []
        for msg in messages:
            role = "user" if msg["role"] == "tool" else msg["role"]
            entry: dict[str, Any] = {"role": role}
            if msg.get("tool_calls"):
                entry["content"] = [
                    {
                        "type": "tool_use",
                        "id": tc["id"],
                        "name": tc["name"],
                        "input": tc["arguments"],
                    }
                    for tc in msg["tool_calls"]
                ]
            elif msg.get("tool_call_id"):
                entry["content"] = [
                    {
                        "type": "tool_result",
                        "tool_use_id": msg["tool_call_id"],
                        "content": str(msg.get("content", "")),
                    }
                ]
            else:
                entry["content"] = msg.get("content", "")
            result.append(entry)
        return result

    def _parse(self, body):
        text_parts, tool_calls = [], []
        for block in body.get("content", []):
            if block["type"] == "text":
                text_parts.append(block["text"])
            elif block["type"] == "tool_use":
                tool_calls.append(
                    {
                        "id": block["id"],
                        "name": block["name"],
                        "arguments": block.get("input", {}),
                    }
                )
        return {
            "role": "assistant",
            "content": "\n".join(text_parts) or None,
            "tool_calls": tool_calls,
            "finish_reason": body.get("stop_reason"),
            "usage": body.get("usage"),
        }


PROVIDERS: dict[str, type[BaseProvider]] = {
    "openai": OpenAIProvider,
    "anthropic": AnthropicProvider,
}


def get_provider(name: str, api_key: Optional[str] = None) -> BaseProvider:
    cls = PROVIDERS.get(name)
    if not cls:
        raise ValueError(f"Unknown provider: {name}")
    return cls(api_key=api_key)
